### PR TITLE
Create Dropbox remote

### LINKED ISSUE #1277: remote: support Dropbox

### PR DESCRIPTION
**Pull request by gmrukwa**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Don't have time to play with that for the next few days, but uploading to make it visible already. Will polish that soon and:

- [x] create docs update
- [x] add tests
- [ ] discuss the app configuration for the API - you guys need to create your own app on Dropbox for the API usage to keep control over it, probably @efiop would be the proper person

Resolves #1277

### CODE REVIEW

**Comment by gmrukwa** [comment]
@efiop, just pinging to let you know the direction this issue is going.

**Comment by efiop** [comment]
Thank you @gmrukwa ! :pray: At the first glance, this looks good. We'll need some time for a proper review though, we currently have a lot on our plate, so thank you in advance for your patience :slightly_smiling_face:

**Comment by gmrukwa** [comment]
Okay guys, I've created docs, tests and the last element to do is to provide the details of the Dropbox API connection - this is to be done by you. Not sure how you want to secure the App Key and App Secret, but [here](https://github.com/iterative/dvc/pull/4793/files#diff-d0bb1fd2cb5f725432045fae5b6d0fc2cffe67b24cff73cc56210379e356401fR29-R30) is the spot.

You need to go to the [Dropbox App Console](https://www.dropbox.com/developers/apps), create your application ([permission and scope details here](https://github.com/iterative/dvc.org/pull/1896/files#diff-6cdebf0ee347687dc2ddd507f881bad8c0c29ab991b856a5ac37e2fc55d36b3cR42-R52)). Dropbox allows you to apply for production usage as soon as you have 50 users linked to your app, but there's some possibility to have the acceptance in the eager mode.

**Comment by shcheklein** [comment]
Okay, will keep track of the thing to review/improve/discuss here. Let's follow up in comments as we address and resolve, and check the box + leave a link to the comment.

- [ ] Check that `download` API in use doesn't fetch the whole file into memory (we had that with GDrive)
- [ ] Check exceptions handling- files not found, general errors. Consult with @efiop on what should we unify. At minimum it's probably better to warp all "expected" exceptions into a DVC-specific one. (check rclone)
- [ ] Support multiple remotes- credentials path variable, similar to GDrive
- [ ] Rate limits - retry? - check rclone.
- [ ] Can we set remote to avoid local sync via API?

File: dvc/tree/dropbox.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    @abstractmethod
+    def can_auth(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def _gather_auth(self):
+        raise NotImplementedError
+
+    def auth(self):
+        import dropbox
```
Comment (shcheklein) [inline dvc/tree/dropbox.py:41]: @efiop do we have a mechanism these days to handle gracefully the failure? (and recommend installing it with a specific dependency)?

File: dvc/tree/dropbox.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+            config = json.load(infile)
+        return {
+            ACCESS_TOKEN: config.get(ACCESS_TOKEN, ""),
+            EXPIRATION: config.get(EXPIRATION, ""),
+            REFRESH_TOKEN: config[REFRESH_TOKEN],
+        }
+
+    def save(self, creds):
+        with open(self._cred_location, "w") as outfile:
+            json.dump(creds, outfile)
```
Comment (shcheklein) [inline dvc/tree/dropbox.py:137]: I'm getting:

```
 File "/Users/ivan/Projects/dvc/dvc/tree/dropbox.py", line 137, in save
    json.dump(creds, outfile)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
```

on the regular CLI flow and a file that is written partially

File: dvc/tree/dropbox.py (hunk 1 above)
Comment (efiop) [reply dvc/tree/dropbox.py:41]: Sure, this needs REQUIRES https://github.com/iterative/dvc/blob/dca48fe8d7f86e49dad856bad1e970b837912793/dvc/tree/s3.py#L24 and the rest will be handled automatically.

Comment (gmrukwa) [reply dvc/tree/dropbox.py:41]: It already is addressed [here](https://github.com/iterative/dvc/pull/4793/files#diff-d0bb1fd2cb5f725432045fae5b6d0fc2cffe67b24cff73cc56210379e356401fR147).

File: dvc/tree/dropbox.py (hunk 2 above)
Comment (gmrukwa) [reply dvc/tree/dropbox.py:137]: Good catch, thanks! I'm on it 🙂

**Comment by gmrukwa** [comment]
I see that you merged some changes recently that make the build breaking, going to check what's up...

**Comment by gmrukwa** [comment]
There's an issue with dependencies for S3 mocks in Python < 3.8 after rebasing onto upstream changes.

**Comment by shcheklein** [comment]
Regarding the rate limits: https://www.dropbox.com/lp/developers/reference/dbx-performance-guide

It looks like we need to test on parallel upload of many files (that's what DVC does by default anyway). It might be very unstable if we don't handle 429 properly and do not respect Retry-after.

We can do parallel uploads, but it might require changing the remote class interface.

**Comment by pared** [comment]
@gmrukwa could you rebase? Dependencies problems should be fixed on the current master.

**Comment by shcheklein** [comment]
Testing it on a bunch of images. This is extremely slow for me for now- is it related to the rate limits I shared above?

Screen Shot 2020-11-14 at 12 34 45 PM

Again, we can compare with rclone, I think.

**Comment by shcheklein** [comment]
Getting an exception now on uploading a dir (it got interrupted in the middle) . Now I can't recover from this.

```
ERROR: failed to upload '../../../../tmp/cache/f7/a911d2e8b3665d3656107a1e054e8f' to 'dropbox://my-project/f7/a911d2e8b3665d3656107a1e054e8f' - ApiError('00d392019d75463ba17df2870ee166ee', UploadSessionFinishError('lookup_failed', UploadSessionLookupError('incorrect_offset', UploadSessionOffsetError(correct_offset=6007346))))
ERROR: failed to upload '../../../../tmp/cache/05/be03208df9c660e9bc0e4752c221ad.dir' to 'dropbox://my-project/05/be03208df9c660e9bc0e4752c221ad.dir'
ERROR: failed to push data to the cloud - 2 files failed to upload
```

**Comment by shcheklein** [comment]
@gmrukwa hey! could you please take a look at those issue we are getting?

**Comment by gmrukwa** [comment]
@shcheklein absolutely - I just need to sort few more things out and hopefully today I will be able to work on that.

**Comment by gmrukwa** [comment]
@shcheklein, I do have everything according to the guide, especially that they embedded most of these considerations into the SDK. I've checked and the upload is done in very similar manner in rclone ([large file upload](https://github.com/rclone/rclone/blob/master/backend/dropbox/dropbox.go#L1311-L1413), [small vs large file upload switch](https://github.com/rclone/rclone/blob/master/backend/dropbox/dropbox.go#L1436-L1447)). The only part I'm missing is retry they have implemented. I've prepared a test for the scenario you described, but it's probably a standard issue with the Dropbox API.

On the other hand, Dropbox has some problems when running as desktop client with tons of small files a well.

**Comment by shcheklein** [comment]
@gmrukwa thanks for the udpate!

> Could you provide some more details on how you ended up in a non-recoverable state?

I was interrupting it with Ctrl-C during the upload process of many image files as far as I remember.

> For tons of small files the upload speed is at the level of 300-500 KB / s

per file? or overall? have you tried to run rclone - is it the same?

Dropbox suggests to use `Batch Upload` in that case. I'm curious if rclone utilizes it? How hard would it be to try it? Probably not an easy change.

**Comment by gmrukwa** [comment]
> I was interrupting it with Ctrl-C during the upload process of many image files as far as I remember.

The test I prepared uses process killing. Not sure yet, how I could simulate Ctrl-C more precisely here.

> per file? or overall? have you tried to run rclone - is it the same?

Unfortunately, it's overall speed. Will test that with rclone as well. But you need to keep in mind that it was tested with 5KB and 100KB files. I remember seeing similar performance of the Dropbox client itself when handling numerous small files. That's not a great news, so I'll test once again with rclone. The only promising option I see (that we discussed already) is extended below.

> Dropbox suggests to use Batch Upload in that case. I'm curious if rclone utilizes it? How hard would it be to try it? Probably not an easy change.

From the technical perspective, there are two aspects to consider.

1. `Batch Upload` is not actually a batch upload of multiple files, but a batch commit of multiple files after they get uploaded. That _could_ potentially be useful for the case with numerous small files _if_ the issue is with _committing_ the content, not with the upload. Luckily, that seems quite likely.
2. Current DVC `BaseTree` API exhibits only methods that upload a single file (there's `upload` vs `_download_file` and `_download_dir`). I am afraid that introducing `Batch Upload` mechanism here would require breaking changes all over the codebase. That's something I would not be able to handle alone for sure, especially not until the end of the year. Additionally I doubt you want to introduce so much change just for introducing a single remote.

**Comment by shcheklein** [comment]
> The test I prepared uses process killing. Not sure yet, how I could simulate Ctrl-C more precisely here.

that's the problem. Can't reproduce it now with killing the process :(

Let's wait for it to appear again, I guess as we test more.

> I'd decide though, whether we want to keep that feature enabled for Dropbox, as the folders are not uniquely identified by IDs

yep, def not needed as part of this PR. Just check that we handle it gracefully. We can figure out this later.

> Unfortunately, it's overall speed. 

that's sad :(. Let's see if rclone does a better job here. 

> I am afraid that introducing Batch Upload mechanism here would require breaking changes all over the codebase.

I hope not. But it'll require some refactoring. I would expect that we will need to override or wrap `_process` from the `remote/base.py`. Not sure about the details- first need to take a look at the batch API itself. I also that this function is extremely coupled and does too many things- some split might be required.

On the other hand 300-400 KB/s - would it be even enough for your case? Can we consider this as a reasonable implementation? 

> Azure doesn't BTW

yep, Azure is very outdated. GDrive is a better example.

> You can use none of them, one or both toghether. I think that making DVC targets by default disabled from "selective sync" would make the most sense.

but neither of them could be enabled via API, right? and selective sync is probably a premium feature (which might be fine, since 2GB free account is not very reasonable).

**Comment by gmrukwa** [comment]
> that's sad :(. Let's see if rclone does a better job here.

Just tested `rclone` - it is 1/6 faster, but tested a bit different scenario, when they could actually lock whole directory at once. They scored ~49 secs for 100 files of 100 KB instead of ~63 secs with current implementation of this PR.

> On the other hand 300-400 KB/s - would it be even enough for your case? Can we consider this as a reasonable implementation?

I didn't work with numerous small files, I have few bigger files (some of them around 1 GB). That would be perfectly fine, as the upload speed would be approx. 12 MB / s as per my benchmark today.

> but neither of them could be enabled via API, right? and selective sync is probably a premium feature (which might be fine, since 2GB free account is not very reasonable).

Selective sync can be pre-enabled by API for Teams plan - that's the only opportunity. Anyone can disable sync of a directory in their client manually though.

**Comment by shcheklein** [comment]
> Here the problem is the "Batch API" isn't formed by Dropbox. 

you mean that you have to start "batch session", which means that you have to save it somewhere potentially, etc, etc? Yes, it might complicate things, but it can be part of the internal Dropbox-specific implementation, no need for a general API here. 

we can split files into batches depending on their size (e.g. 1-2GB per batch?). In this case we can potentially get a reasonable performance with multiple jobs (`--jobs`) and it's fine to fail, it will restart some 1-2GB batches. 

Even if it is the case, I guess it won't be part of this PR. It's just good to do a bit of research if we are missing some low hanging fruits.

> Selective sync can be pre-enabled by API for Teams plan - that's the only opportunity. Anyone can disable sync of a directory in their client manually though (including free users with 2 GB of space).

okay. Let's recommend in the docs at least that sync should be disabled one way or another to avoid duplicating data and traffic?

> Just tested rclone - it is 1/6 faster

:(  

Btw, I see that rclone improved performance in some cases by providing a better chunk size, do we do the same? https://github.com/rclone/rclone/issues/103 

-----------

@shcheklein, I would need info from you what I need to adjust here to have it merged. I am now wrapping exceptions, but if you see anything else, please let me know.


yep, I don't see that much left for this PR. It's a very good and solid start. 

One last thing that we'll need to make sure that we set properly are those vars

```
    DEFAULT_VERIFY = False
    LIST_OBJECT_PAGE_SIZE = 1000
    TRAVERSE_WEIGHT_MULTIPLIER = 5
    TRAVERSE_PREFIX_LEN = 3
    TRAVERSE_THRESHOLD_SIZE = 500000
    CAN_TRAVERSE = True
```

from the `base.py`. @pmrowla could you help us with this? what information do you need for this? Dropbox is similar to GDrive - it's API based, APIs are expensive (and potentially limited), so what you be the recommended setup here?

**Comment by gmrukwa** [comment]
> you mean that you have to start "batch session", which means that you have to save it somewhere potentially, etc, etc?

The scenario in `rclone`, on Stack Overflow and everywhere else I saw was as follows:

1. Check file size vs chunk size.
2. If file size < chunk size, use `client.files_upload` and you're done, otherwise proceed with remaining points.
3. Open session `client.files_upload_session_start`.
4. Make cursor for the session `dropbox.files.UploadSessionCursor`.
5. Make commit info for the file that gets uploaded `dropbox.files.CommitInfo`.
6. Upload chunk-by-chunk `client.files_upload_session_append_v2`.
7. Finish session with cursor & commit info `client.files_upload_session_finish`.

No one ever shows how to use the batch mode. ☝️ There are batch operations in the API for moves, deletions, folder creations, but not upload. What I understood from the docs, client must acquire a lock over the namespace (==directory) before committing the file. This is automated for `client.files_upload` within the SDK. There exist functions:

- [client.files_lock_file_batch](https://dropbox-sdk-python.readthedocs.io/en/latest/api/dropbox.html#dropbox.dropbox.Dropbox.files_lock_file_batch)
- [client.files_upload_session_finish_batch](https://dropbox-sdk-python.readthedocs.io/en/latest/api/dropbox.html?highlight=files_upload#dropbox.dropbox.Dropbox.files_upload_session_finish_batch)

Perhaps we could acquire locks in batch mode for a single client and then commit everything at once. Not sure though how to use these locks, as the SDK doesn't accept them while committing.

> we can split files into batches depending on their size (e.g. 1-2GB per batch?). In this case we can potentially get a reasonable performance with multiple jobs (--jobs) and it's fine to fail, it will restart some 1-2GB batches.

Yeah, I can enable that. Even though `rclone` claims that it can speed up upload max by 10%. And this makes a difference only for big files - for small ones we're using `client.files_upload` anyway. The chunk cannot exceed 150 MB and needs to be a multiple of 4 MB as stated in the docs. `rclone` has default of 48 MB chunk.

> okay. Let's recommend in the docs at least that sync should be disabled one way or another to avoid duplicating data and traffic?

On it.

To summarize the next steps:

- [ ] wrap exceptions
- [ ] parameterize chunk size
- [ ] recommend enabling selective sync / smart sync in the docs
- [ ] adjust the above mentioned variables

**Comment by gmrukwa** [comment]
@shcheklein, I tried to address all the above points and it seems I've finished. Could you take a look now? If we're done, let's merge it - my migration from Azure awaits. 😂

**Comment by shcheklein** [comment]
@gmrukwa thanks a lot 🙏 ! I'll try to take a look asap!

**Comment by gmrukwa** [comment]
Rebased to resolve merge conflicts ☝🏻

File: dvc/tree/dropbox.py [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        }
+
+    def save(self, creds):
+        _creds = creds.copy()
+        _creds[EXPIRATION] = _creds[EXPIRATION].isoformat()
+        with open(self._cred_location, "w") as outfile:
+            json.dump(_creds, outfile)
+
+
+def path_info_to_dropbox_path(path_info):
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:236]: I've updated the code in https://github.com/gmrukwa/dvc/pull/1 to allow empty hostname so that we don't have to deal with `bucket` that is not bucket at all (and to deal with cases like `import-url dropbox:///file.pdf`

still, I wonder what is the best way to handle this leading `/` here, @efiop ? Dropbox expects them this way? Is there a convenience method in PathInfo to return path with `/`?

**Review by shcheklein — CHANGES_REQUESTED** [review]
Please see some requests for refactoring, see PR I made to wire up properly import-ur/get-ulr/external deps and fix certain things

Next steps on my end:

- [x] Try to figure out why is it SOOOO slow. It took me ~10minutes to `import-url` 140Mb/32files. Is it expected?
- [ ] I got `my-project (Ignored Item Conflict) (Ignored Item Conflict)` in my Dropbox - what is this?
- [ ] figure out case sensitivity `- /Users/ivan/Projects/test-dropbox/Books/../books/` ... should we use regular path, not lower in `walk_files`?
- [ ] review tests and try to run them, wire up import url and other regular tests for remotes
- [ ] check if APP is setup properly
- [ ] review docs once again

File: dvc/tree/dropbox.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+    @cached_property
+    def client(self):
+        cred_prv = EnvCredProvider()
+        if cred_prv.can_auth():
+            logger.debug("Logging into Dropbox with environment variables")
+            return cred_prv.auth().client
+        cred_prv = FileCredProvider(self.repo)
+        if cred_prv.can_auth():
+            logger.debug("Logging into Dropbox with credentials file")
+            return cred_prv.auth().client
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:278]: I didn't spot the refresh token to change anytime. Therefore, actually you only need the refresh token to get the access token each time before the API call. The remaining elements could probably be dropped, but I introduced them to keep compatibility if they start to timeout the refresh token.

File: dvc/tree/dropbox.py [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+            REFRESH_TOKEN: auth.refresh_token,
+        }
+
+    def save(self, creds):
+        pass
+
+
+class EnvCredProvider(DropboxCredProvider):
+    def can_auth(self):
+        return bool(os.environ.get(REFRESH_TOKEN, False))
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:189]: I'll refer to my comment above. You actually need only the refresh token, as it allows you to get the access token anytime. Refresh token is not ephemeral one-time token just for a single refresh, you can use it as many time as you want.

File: dvc/tree/dropbox.py [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+    def __init__(self, repo):
+        self._cred_location = os.environ.get(
+            self.DROPBOX_CREDENTIALS_FILE,
+            os.path.join(repo.tmp_dir, self.DEFAULT_FILE),
+        )
+
+    def can_auth(self):
+        if not os.path.exists(self._cred_location):
+            return False
+        with open(self._cred_location) as infile:
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:217]: True, we could.

File: dvc/tree/dropbox.py [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+                        )
+                        cursor.offset = wrapped.tell()
+
+    def _download(
+        self, from_info, to_file, name=None, no_progress_bar=False, **_kwargs
+    ):
+        import dropbox
+
+        from_path = path_info_to_dropbox_path(from_info)
+        logger.debug("Downloading " + from_path + " to " + to_file)
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:439]: True.

File: dvc/tree/dropbox.py [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+                    "use the `dropbox_user_credentials_file` config\n"
+                    "   option if you use multiple Dropbox remotes with "
+                    "different email accounts.\n\nDetails".format(
+                        from_path, FileCredProvider.DEFAULT_FILE
+                    )
+                ) from ex
+            if ex.error.is_unsupported_file():
+                raise DvcException(
+                    "Path '{}' is not downloadable:\n\n"
+                    "Confirm the file is a casual file, not a Dropbox thing."
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:458]: They do have e.g. Dropbox Paper documents or Dropbox Vault items which are not downloadable this way.

File: dvc/tree/dropbox.py [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+            # expires in 4 hrs
+            return self.client.files_get_temporary_link(path).link
+        except dropbox.exceptions.ApiError as ex:
+            if ex.error.is_path() and ex.error.get_path().is_not_found():
+                raise DvcException(
+                    "Path not found for '{}':\n\n"
+                    "1. Confirm the file exists and you can access it.\n"
+                    "2. Make sure that credentials in '{}'\n"
+                    "   are correct for this remote e.g. "
+                    "use the `dropbox_user_credentials_file` config\n"
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:299]: It is, not a setting, but an env variable.

File: dvc/tree/dropbox.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:236]: Dropbox generally expects a path that is like `/a/path/to/file`, which gets rooted at the root of your Dropbox home. I couldn't get it to work simply without such a construct unfortunately. You will rather know some better way.

File: dvc/tree/dropbox.py [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+    # returned in some cases.
+    LIST_OBJECT_PAGE_SIZE = 1000
+    # Always prefer traverse for Dropbox since API usage quotas are a concern.
+    TRAVERSE_WEIGHT_MULTIPLIER = 1
+    TRAVERSE_PREFIX_LEN = 2
+
+    def __init__(self, repo, config):
+        super().__init__(repo, config)
+
+        self.path_info = self.PATH_CLS(config.get("url", "dropbox://default"))
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:258]: Right, it should be None and potentially throw an exception.

File: dvc/tree/dropbox.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:236]: yep, I did some changes to support URL without hostname (which is not present here). Usually the look like `file:///path` (mind triple `///`).

File: dvc/tree/dropbox.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:458]: yep, makes sense, `Dropbox thing` sounds a bit strange though :), how do they call them in docs?

File: dvc/tree/dropbox.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:189]: interesting, thanks! 

Thinking from the user perspective I still think it's better try to this type of auth if we detected any of those vars (and fail if they are not enough, e.g. REFRESH_TOKEN is not set). Just to be extra explicit. To avoid situations you set some vars, but it still uses file w/o any explanation. WDYT?

File: dvc/tree/dropbox.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:278]: yep, but it means we are going to keep refreshing them every run? (extra API call, extra time to an operation). Probably not the biggest problem, but I thought it can be a simple fix since you already have `save()` mechanics.

[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:278]: The access token lives 4 hours. It's not much. We create the client once per dvc command being ran, so the update takes place at most once per CLI run.

File: dvc/tree/dropbox.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:189]: I thought of two approaches:

- requiring just the refresh token - that seems reasonable if it's the only must-have element
- requiring everything - in the case they will shorten life of refresh token or make it usable just once

It's your decision, I can adjust the code as needed.

File: dvc/tree/dropbox.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (gmrukwa) [reply dvc/tree/dropbox.py:458]: They do not call them in the SDK docs at all (or at least I haven't spotted) 😂 There's just written, that something may not be downloadable.

[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:458]: Look like we call them Non-downloadable files - https://www.dropbox.com/lp/developers/reference/dbx-file-access-guide#files-that-require-special-consideration .

`Confirm that it's a regular file, not a non-downloadable Dropbox file`

(I don't like it either, but at least it's sounds a bit more formal and can be googled)

File: dvc/tree/dropbox.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:189]: So, there are two different things (correct me if I wrong)

1. Trigger - some factor that defines that we go the path of a specific auth (and don't try others)
2. Validation - check if set of params is enough to go that path

So, in this specific case (to my mind):

1. Can be any of the env vars present. It's better from the user perspective, we are extra explicit.
2. Can check existence of the refresh token. Or also allow both keys to be present (I'm not sure if it makes sense at all)?


Btw, how would the actual workflow with env vars would look like? Where do users get them in the first place?

File: dvc/tree/dropbox.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:299]: yep, but message looks strange (comes from the GDrive support which supports `user_credentials_file` on the config file level? and which should be the case here as well, but later, after this PR I guess ... or I can add in my PR, don't focus on this too much

[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/tree/dropbox.py:299]: This long message is still massive, and has a lot of duplicates, would be great to dry it a bit

File: dvc/tree/dropbox.py [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+    @_dropbox_retry
+    def files_upload_session_finish(self, *args, **kwargs):
+        return self.client.files_upload_session_finish(*args, **kwargs)
+
+    @_dropbox_retry
+    def files_download(self, *args, **kwargs):
+        return self.client.files_download(*args, **kwargs)
+
+
+ACCESS_TOKEN = "DROPBOX_ACCESS_TOKEN"
```
Comment (shcheklein) [inline dvc/tree/dropbox.py:94]: It looks like rclone uses a different way of authenticating? It gets ` 

`{"access_token":".***","token_type":"bearer","expiry":"0001-01-01T00:00:00Z"}`

I don't see a refresh token, and expiry is not set. Do you know something about this?

**Comment by shcheklein** [comment]
@gmrukwa the biggest problem for me right now is that download takes a lot of time for a directory of ~142MB which contains files ~1MB to 10MB. 

`rclone` does it in 20s
`dvc pull` works minutes, sometimes a single file is slower than 20s

Do you have any idea what to try, where to look for the problem?

File: dvc/tree/dropbox.py (hunk 11 above)
Comment (gmrukwa) [reply dvc/tree/dropbox.py:94]: They may be using legacy auth, which will be dropped soon by Dropbox.

**Comment by gmrukwa** [comment]
> @gmrukwa the biggest problem for me right now is that download takes a lot of time for a directory of ~142MB which contains files ~1MB to 10MB.
> 
> `rclone` does it in 20s
> `dvc pull` works minutes, sometimes a single file is slower than 20s
> 
> Do you have any idea what to try, where to look for the problem?

They use `download` function from SDK as well ([here](https://github.com/rclone/rclone/blob/master/backend/dropbox/dropbox.go#L1280-L1319)), here we do as well. I do not have any other ideas for that.

**Comment by shcheklein** [comment]
@gmrukwa yep, I have no idea either. Let me try to debug it a bit during this weekend to see if I can find something.

**Comment by shcheklein** [comment]
@gmrukwa improved download tremendously (it's now 25Mb/s vs 150Kb/s on a large file, and downloads a dir with files faster than rclone) here https://github.com/gmrukwa/dvc/pull/1/commits/3ea0908c972b8a3e928c1cff502be5809c889566 . It's a simple fix. Haven't had time to check/compare upload yet.

**Comment by gmrukwa** [comment]
@shcheklein, that fix makes perfect sense! I was looking for that option in the Dropbox SDK actually, but it didn't came to my mind that download chunk size is steered on the level of `iter_content`... 🤯

**Comment by shcheklein** [comment]
@gmrukwa 

> I'd use the chunk_size setting value here, wouldn't it make sense?

good question 🤔 to be honest, not sure about this. My only concern is if they have a bit different limits and one expects a range in KBs, second in MBs to have a good performance. It looks like rclone's setting doesn't affect download?

**Comment by gmrukwa** [comment]
> @gmrukwa
> 
> > I'd use the chunk_size setting value here, wouldn't it make sense?
> 
> good question 🤔 to be honest, not sure about this. My only concern is if they have a bit different limits and one expects a range in KBs, second in MBs to have a good performance. It looks like rclone's setting doesn't affect download?

rclone seems to be using different method from the SDK, more like `files_download_to_file`, but with DVC we want to show the download progress. I've reviewed the code for `files_download_to_file` and Dropbox internally indeed uses `2**16` chunk size for download. Let's leave as you proposed then.

**Comment by efiop** [comment]
Quick update: we are migrating to fsspec and that will be our plugin mechanism for the future. We won't be able to provide proper support for dropbox ourselves, so the best way to go about it would be for you to implement dropboxfs using fsspec and maintain that implementation (try registering it in fsspec, other people might find it useful too). On our end, we will allow using that as a plugin as seamlessly as possible. Thank you so much for contributing!

There is an existing 3rd party implementation https://github.com/MarineChap/dropboxdrivefs , but I'm not sure how stable it is and whether it supports the same features that this PR supports. There has been some pretty active discussion around it in https://github.com/intake/filesystem_spec/pull/207 . We might want to consider either contributing to the existing package or creating a new implementation.
